### PR TITLE
Add homepage url to user profile

### PR DESCRIPTION
**Pull request by jacklynhma**
Objective: 
- Add a homepage url in the profile 

Screenshot 2024-11-14 at 17 59 50

According to the tests the twitter and the homepage URL should display here but I have not been able to display this on main (?): 
Screenshot 2024-11-14 at 18 01 51

I was told that the icon will show on production: 
Screenshot 2024-11-14 at 18 03 36

### CODE REVIEW

File: app/models/user.rb [hunk 1]
```diff
@@ -82,6 +82,8 @@ class User < ApplicationRecord
     message: "can only contain letters, numbers, and underscores"
   }, allow_nil: true, length: { within: 0..20 }
 
+  validates_formatting_of :homepage_url, using: :url, allow_blank: true
```
Comment (jacklynhma) [inline app/models/user.rb:85]: When I do `allow_nil: true` twitter tests in the profile_test.rb would fail. Seeking advice if this is best practice.

Comment (segiddins) [reply app/models/user.rb:85]: ah it's because homepage_url is optional, but the form always submits an empty string. I forget if there's a good way to handle this

File: app/views/dashboards/_subject.html.erb [hunk 2]
```diff
@@ -23,6 +23,17 @@
   </div>
 <% end %>
 
+<% if user.homepage_url.present? %>
+  <div class="flex items-center mb-4 text-b3 lg:text-b2">
+    <%= icon_tag("link", color: :primary, class: "w-6 text-orange mr-3") %>
+    <p class="text-neutral-800 dark:text-white"><%=
+      link_to(
```
Comment (segiddins) [inline app/views/dashboards/_subject.html.erb:30]: note to self: suggest the appropriate `rel=none; noreferrer` options

**Review by segiddins — APPROVED** [review]
don't worry about the coverage change

File: app/views/dashboards/_subject.html.erb (hunk 2 above)
Comment (martinemde) [reply app/views/dashboards/_subject.html.erb:30]: looks like `rel="nofollow"` is what github uses on their profile links. I'm agnostic about "noreferrer" in addition.

File: app/models/user.rb (hunk 1 above)
Comment (martinemde) [reply app/models/user.rb:86]: I think this is expected.

File: app/views/dashboards/_subject.html.erb [hunk 3]
```diff
@@ -23,6 +23,19 @@
...
+        user.homepage_url,
+        user.homepage_url,
+        rel: "nofollow"
```
Comment (jacklynhma) [inline app/views/dashboards/_subject.html.erb:33]: opps sorry I didn't realize there was a final decision on the `rel="nofollow"`.  Thank you for adding these changes :) @martinemde

File: app/views/profiles/show.html.erb [hunk 4]
```diff
@@ -94,6 +94,18 @@
             )
           %>
         <% end %>
+        <% if @user.homepage_url.present? %>
+          <p id="homepage-url">
+            <%=
+              link_to(
+                @user.homepage_url,
+                @user.homepage_url,
```
[the remark this answers is not in the document]
Comment (martinemde) [reply app/views/profiles/show.html.erb:102]: Maybe we should consider something like HackerOne's "you're about to leave this site for ...". Github appends `http://` to urls that don't have either http / https in the front, and they are probably doing more.

**Comment by martinemde** [comment]
Blocking temporarily while we make sure we're sanitizing the urls. I suspect that since we already allow urls from gems, this isn't a whole lot worse, but I want to double check.

**Comment by jacklynhma** [comment]
@martinemde Thanks for calling this out. 

I could be wrong, but while I was looking at the code, I did not see any sanitizing for the URL. 

There are some safeguards though. 
- If I comment out this new line of code https://github.com/rubygems/rubygems.org/pull/5240/files#diff-9802ca3c9c4cf89904fd44bc114e35ebdf2c5dd3d5b645491e2b253e1afef29bR357 
- Add `javascript:alert('XSS');` to the homepage_url field in the edit profile 
- Navigate to `/profiles/(user_name)` 
- Try to click on the homepage_url 
- I will see the below message 
Screenshot 2024-11-21 at 16 55 02

which is tied to https://github.com/rubygems/rubygems.org/blob/master/config/initializers/content_security_policy.rb#L33 and which protects against XXS attacks. 

However with this validation https://github.com/rubygems/rubygems.org/pull/5240/files#diff-9802ca3c9c4cf89904fd44bc114e35ebdf2c5dd3d5b645491e2b253e1afef29bR357  
It looks like the code does prevent `javascript:alert('XSS');` from even being submitted 
https://github.com/mdespuits/validates_formatting_of/blob/664b7c8b1ae8c9016549944fc833737c74f1d752/lib/validates_formatting_of/method.rb#L19
So something like this will be caught and the below error message will show: 
Screenshot 2024-11-21 at 16 33 36


What we can also do is  with that `sanitize` method 
```
      sanitize( link_to(
         user.homepage_url,
         user.homepage_url,
        rel: "nofollow"
      ), tags: %w(a), attributes: %w(href rel))
```
And then it will remove the href from the link and make it unclickable. 
Screenshot 2024-11-21 at 17 18 30

But I understand that true sanitizing would remove everything we don't want in the string. I could also look into this. Please let me know how you would like me to proceed or if I am completely off the mark.

File: test/models/user_test.rb [hunk 5]
```diff
@@ -172,6 +172,24 @@ class UserTest < ActiveSupport::TestCase
...
+      should_not allow_value("<html><body>hi</body><html>").for(:homepage_url)
+      should_not allow_value("javascript:alert('hello');").for(:homepage_url)
+      should_not allow_value("file:///etc/passwd").for(:homepage_url)
+      should_not allow_value("file://C:/Windows/System32/cmd.exe").for(:homepage_url)
+      should_not allow_value("data:text/html,<script>alert('XSS')</script>").for(:homepage_url)
+      should_not allow_value("data:text/html;base64,SGVsbG8sIFdvcmxkIQ==").for(:homepage_url)
+      should_not allow_value("data:text/html,<img src='x' onerror='alert(1)'>").for(:homepage_url)
+      should_not allow_value("data:text/html,<iframe src='javascript:alert(1)'></iframe>").for(:homepage_url)
+      # should_not allow_value("http://www.site.site?#redirect=www.fake-target.site").for(:homepage_url)
+      # should_not allow_value("http://www.site.com/redirect?url=http://www.malicious-site.com").for(:homepage_url)
```
Comment (jacklynhma) [inline test/models/user_test.rb:187]: I am guessing this is a yes, but I commented this out because these tests will fail, and I wanted to clarify whether these are edge cases we want to cover.

File: lib/http_url_validator.rb [hunk 6]
```diff
@@ -0,0 +1,10 @@
+class HttpUrlValidator < ActiveModel::EachValidator
```
Comment (jacklynhma) [inline lib/http_url_validator.rb:1]: I read about how the URI validation occurs in the Ruby gem(https://github.com/rubygems/rubygems/blob/master/lib/rubygems/specification_policy.rb#L450-L459)  and tried to mimic it since I figured we would want the same validations to happen in both places. 

I extracted this validation into a custom validator because I knew that other URLs would eventually be added to the user profile, and we could establish a consistent way of validating the URLs. 

I just also wanted to call out that although the matching pattern `validUriPattern = %r{\Ahttps?://([^\s:@]+:[^\s:@]*@)?[A-Za-z\d\-]+(\.[A-Za-z\d\-]+)+\.?(:\d{1,5})?([\/?]\S*)?\z} # :nodoc:` could have been achieved with the `validate_the_formatting_of`  a custom validation will allow us to add checks if we want to extract params such as `redirect`

File: app/views/dashboards/_subject.html.erb [hunk 7]
```diff
@@ -21,6 +21,19 @@
     </div>
   <% end %>
 
+<% if user.homepage_url.present? %>
+  <div class="flex items-center mb-4 text-b3 lg:text-b2">
+    <%= icon_tag("link", color: :primary, class: "w-6 text-orange mr-3") %>
+    <p class="text-neutral-800 dark:text-white"><%=
+      link_to(
+        truncate(user.homepage_url, length: 20),
```
Comment (jacklynhma) [inline app/views/dashboards/_subject.html.erb:29]: Taking into account really long homepage URLs. 
I chose length 20 because I saw that is the length we limit our Twitter username 

### Before: 
Screenshot 2024-11-21 at 16 33 36


### After 
Screenshot 2024-12-03 at 13 54 54
Screenshot 2024-12-03 at 13 55 07

File: app/views/profiles/show.html.erb (hunk 4 above)
[the remark this answers is not in the document]
Comment (jacklynhma) [reply app/views/profiles/show.html.erb:102]: Hi @martinemde 
 I added the confirmation dialog here before the user gets redirected:  https://github.com/rubygems/rubygems.org/pull/5240/commits/50441e67f2dccbbe4c47430b047607007f1e6be1

For your second suggestion, I am seeking clarity on your ask to append `http://` 

Are these use cases correct: 
| User Enters in homepage URL  | Expected outcome |
| ------------- | ------------- |
| `www.userhomepage.com`  | `http://www.userhomepage.com`   |
| `javascript:alert('hello')` | `http://javascript:alert('hello')`  |
| `file:///etc/passwd` | `http://file:///etc/passwd`| 
| `http://javascript:alert('hello')`| `http://javascript:alert('hello')`| 
 | `http://www.userhomepage.com`   | `http://www.userhomepage.com`   |
 | `https://www.userhomepage.com`   |  `https://www.userhomepage.com`   |

If so, that means there will be no error messages and the user's input will persist in the database with the appended `http://`

**Comment by jacklynhma** [comment]
https://github.com/rubygems/rubygems.org/pull/5240#discussion_r1868323105
https://github.com/rubygems/rubygems.org/pull/5240#issuecomment-2495130324
@martinemde After rereading your above messages, I think below are the use cases you are expecting. Please let me know and I can update the code. Thank you for your feedback. 

User Enters in homepage URL | Expected outcome
-- | --
www.userhomepage.com | http://www.userhomepage.com
javascript:alert('hello') | Error message rendered, record is not updated
file:///etc/passwd | Error message rendered, record is not updated
http://javascript:alert('hello') |  Error message rendered, record is not updated
http://www.userhomepage.com | http://www.userhomepage.com
https://www.userhomepage.com | https://www.userhomepage.com

File: app/views/profiles/show.html.erb (hunk 4 above)
[the remark this answers is not in the document]
Comment (martinemde) [reply app/views/profiles/show.html.erb:102]: I wonder if wrapping this in `h()` would silence the warning. It should have no effect for safe urls.

**Comment by martinemde** [comment]
I think these are being flagged "correctly", in that it's a valid concern. Ultimately it could be that we're validating the inputs correctly but we should take whatever care is necessary to also convince the security bots that we are not doing something dumb.

File: app/views/profiles/show.html.erb (hunk 4 above)
[the remark this answers is not in the document]
Comment (martinemde) [reply app/views/profiles/show.html.erb]: Sorry, I didn't answer you http question. I think requiring that the url start with http(s) as a validation is better than prepending it.

**Comment by jacklynhma** [comment]
@martinemde your suggestion to add `h()` worked! 🥳  I am noticing the lint/brakeman build failing but I do not know how to access the `reports/brakeman/brakeman.json` file to see the actual error.

File: test/system/authorizing_profile_update_test.rb [hunk 8]
```diff
@@ -1,7 +1,7 @@
 require "application_system_test_case"
 require "test_helper"
 
-class ProfileTest < ApplicationSystemTestCase
```
Comment (jacklynhma) [inline test/system/authorizing_profile_update_test.rb:4]: When running tests locally, I receive the below error: Screenshot 2024-12-13 at 21 56 42

I realize this is a bug I created in a previous PR since I used the same class name. Changing the file to another name allowed the tests to run.

**Comment by jacklynhma** [comment]
@martinemde I was able to get the brakeman build to pass  by prepending `https://` to the urls just in case in the future the model changes by accident and adding your suggestion to add `h()` 🥳 

This build test failed but it passes locally ..I do not think it might be related to my PR (?) 
Screenshot 2024-12-15 at 13 14 10

File: test/system/authorizing_profile_update_test.rb (hunk 8 above)
Comment (jacklynhma) [reply test/system/authorizing_profile_update_test.rb:4]: Moving this to its own PR https://github.com/rubygems/rubygems.org/pull/5366
